dropwizard-hibernate/src/main/java/io/dropwizard/hibernate/UnitOfWorkAspect.java

```diff
@@ -89,14 +89,14 @@ public void beforeStart(@Nullable UnitOfWork unitOfWork) {
             try {
                 configureSession();
                 ManagedSessionContext.bind(session);
-                beginTransaction(unitOfWork, session);
             } catch (Throwable th) {
                 session.close();
                 session = null;
                 ManagedSessionContext.unbind(sessionFactory);
                 throw th;
             }
         }
+        beginTransaction(unitOfWork, session);
```

Comment:
The `beginTransaction` here is no longer guarded by `try {} catch {}`, I think what you might want do is catch exceptions, and if you own the session, close/clean/unbind. e.g.:

```
try {
  beginTransaction(unitOfWork, session);
} catch (Throwable th) {
  if (sessionCreated) {
    // Same logic as the previous catch block, e.g. close/unbind/etc
  }
  throw th;
}

Reply:
Valid point, but the documentation of the `UnitOfWorkAspect` class shows the call sequence, that requires the calls to either hit `UnitOfWorkAspect#afterEnd` or `UnitOfWorkAspect#onError`. There are two call sites: 

- `io.dropwizard.hibernate.UnitOfWorkAwareProxyFactory.create(Class<T>, Class<?>[], Object[])`
- `io.dropwizard.hibernate.UnitOfWorkApplicationListener.UnitOfWorkEventListener.onEvent(RequestEvent)`

The first is pretty straight forward and will call into the `onError` method in a catch block, the latter hooks into the jersey events and exceptions in the jersey process should result in the `ON_EXCEPTION` event being generated and invoking `onError`

Reply:
Awesome, that makes sense.